modules-js/react-fleet/src/form-elements/inputs/MemorableDateInput.tsx

```diff
@@ -80,95 +87,125 @@ export default class MemorableDateInput extends React.Component<Props, State> {
-            day: this.initial.getDate().toString(),
-          }
-        : {
-            year: '',
-            month: '',
-            day: '',
-          },
+        ? dateToFields(this.initial)
+        : { year: '', month: '', day: '' },
     };
   }
 
-  // Determine if date range limits have been declared, and whether a given
-  // date is within that range. If not, this.state.error is set to reflect this.
-  private checkIfDateInRange(date: Date): boolean {
-    const errorString = isDateValid(
+  /**
+   * Helper to call isDateValid with all of our constraint props filled in.
+   */
+  private isDateValid(date: Date): string | null {
+    return dateValidError(
       date,
       this.earliest,
       this.latest,
       this.props.onlyAllowPast,
       this.props.onlyAllowFuture
     );
-
-    this.setState({ error: errorString });
-
-    return errorString.length === 0;
   }
 
-  // If date is within range, update value. Otherwise, clear the value.
-  private updateDate(): void {
-    const { year, month, day } = this.state.fields;
-
-    // When Date is called with more than one arg, result is set to midnight.
-    let date: Date | null = new Date(+year, +month - 1, +day);
-
-    // If supplied date is out of range, clear the date value.
-    if (!this.checkIfDateInRange(date)) {
+  /**
+   * Called when the fields change to update state. Calls the handleDate prop
+   * with either a valid date or null.
+   *
+   * Also updates the lastValidityError state so we can keep error messages
+   * around while the user edits.
+   */
+  private updateFields(fields: Fields) {
+    let date: Date | null;
+    let lastValidityError: string | null;
+
+    if (inputCompleteError(fields)) {
       date = null;
+      // We don't reset the validity error if the user hasn’t typed in a new
+      // date. This lets us show the date constraints while they’re editing.
+      lastValidityError = this.state.lastValidityError;
+    } else {
+      date = fieldsToDate(fields);
+      lastValidityError = this.isDateValid(date);
+
+      if (lastValidityError) {
+        date = null;
+      }
     }
 
-    // Call parent component’s method with the new date value.
-    this.props.handleDate(date);
+    this.setState(
+      {
+        fields,
+        lastValidityError,
+      },
+      () => {
+        this.props.handleDate(date);
+      }
+    );
   }
 
   // For the purposes of gracefully showing/hiding error text, we want to track
   // focus/blur across all three fields in order to treat this component itself
   // as having a focus and a blur state.
   private handleFocus = (): void => {
     clearTimeout(this.timeout);
-
     this.setState({ hasFocus: true });
   };
 
   private handleBlur = (): void => {
-    this.timeout = setTimeout(() => this.setState({ hasFocus: false }), 0);
+    this.timeout = setTimeout(() => {
+      let { fields } = this.state;
+
+      // This has the effect of normalizing dates. So, for example, February
```

Comment:
I’d been thinking that it would be nice to do further work on this component during bash week, specifically basic validity (e.g. if September, date max is 30; allow Feb date max 29 if it’s a leap year), and preventing a user from typing a number with more digits than is possible for a given field - would love to chat about it tomorrow!